commons-jetty/src/main/scala/com/avsystem/commons/jetty/rpc/JValueInputOutput.scala

```diff
@@ -0,0 +1,115 @@
+package com.avsystem.commons
+package jetty.rpc
+
+import java.util.Base64
+
+import com.avsystem.commons.serialization.GenCodec.{ReadFailure, WriteFailure}
+import com.avsystem.commons.serialization.{FieldInput, InputType, ListInput, ObjectInput, _}
+import jawn.ast.{JArray, JBool, JFalse, JNull, JNum, JObject, JString, JTrue, JValue}
+
+object JValueOutput {
+  def write[T: GenCodec](value: T): JValue = {
+    var result: Opt[JValue] = Opt.Empty
+    val output = new JValueOutput(value => result = Opt(value))
+    GenCodec.write(output, value)
+    result.getOrElse(throw new WriteFailure("No value written by GenCodec"))
+  }
+}
+
+final class JValueOutput(private val consumer: JValue => Unit) extends AnyVal with Output {
+  def writeNull() = consumer(JNull)
+  def writeBoolean(boolean: Boolean) = consumer(if (boolean) JTrue else JFalse)
+  def writeString(str: String) = consumer(JString(str))
+  def writeInt(int: Int) = consumer(JNum(int))
+  def writeLong(long: Long) = consumer(JNum(long))
+  def writeDouble(double: Double) = consumer(JNum(double))
+  def writeBinary(binary: Array[Byte]) = consumer(JString(Base64.getEncoder.encodeToString(binary)))
+  def writeList() = new JValueListOutput(consumer)
+  def writeObject() = new JValueObjectOutput(consumer)
+}
+
+final class JValueListOutput(consumer: JArray => Unit) extends ListOutput {
+  private val builder = Array.newBuilder[JValue]
+
+  def writeElement() = new JValueOutput(builder += _)
+  def finish() = consumer(JArray(builder.result()))
+}
+
+final class JValueObjectOutput(consumer: JObject => Unit) extends ObjectOutput {
+  private val builder = MMap.newBuilder[String, JValue]
```

Comment:
Might be worth it to use `MMap` implementation which preserves field order. Some `GenCodec` goodies rely on that.